ur_ros2_control_demos/launch/ur_ros2_control.launch.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) 2021 PickNik, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from ament_index_python.packages import get_package_prefix, get_package_share_directory
+
+from launch import LaunchDescription
+from launch.actions import DeclareLaunchArgument, LogInfo
+from launch.substitutions import Command, LaunchConfiguration, PathJoinSubstitution
+
+from launch_ros.actions import Node
+
+import xacro
+
+
+def load_file(package_name, file_path):
+    package_path = get_package_share_directory(package_name)
+    absolute_file_path = os.path.join(package_path, file_path)
+
+    try:
+        with open(absolute_file_path, 'r') as file:
+            return file.read()
+    except EnvironmentError:  # parent of IOError, OSError *and* WindowsError where available
+        return None
+
+
+def generate_launch_description():
+    declared_arguments = []
+    substitutions = []
+    declared_arguments.append(DeclareLaunchArgument(
+        'ur_type',
+        description='Type/series of used UR robot.'))
```

Comment:
A data point on UX: traditionally, in ROS-Industrial, we've not used parameterisation like this as the feedback we received was that "parameters everywhere" actually make things much more complex.

Cognitive overload with ROS is very easy to achieve and I don't see ROS 2 improving on this.

Questions like "is my robot supported?" now require a (potential) user to grok *where* to look, be able to parse the logic and then be able to decipher which parameter values are legal. Compare this to a minor duplication of files which contain the name of the supported robot in their filename.

Counterarguments typically heard are "yes, but we can document the allowed values" and "we'll need to write clear documentation". That's true, but it increases maintenance overhead (both initially and it needs to be kept up-to-date) and is again another source of information.

Reply:
My Counter argument. We will create a launch file for each robot type. This "underlying" file is not intended for use directly, especially not by beginners. I just want to save multiple hundreds of lines of code by reusing launch files.

Other than that, there is a possibility to define legal values. I will add those also.

Reply:
Approach with parametrized launch file as underlying base sounds good for me even for newbie.